aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -2,331 +2,447 @@
         }
     }
 
-    /// Applies self on top of previous change, merging them together. Note
-    /// that the strict ordering here is crucial for catching overflows correctly.
-    pub fn merge_with_previous_aggregator_change(
-        &mut self,
-        previous_change: &AggregatorChange,
-    ) -> PartialVMResult<()> {
-        match self {
-            // If the current state is Data, then merging with previous state won't change anything.
-            AggregatorChange::Data { .. } | AggregatorChange::StringData { .. } => Ok(()),
-            AggregatorChange::AggregatorDelta {
-                delta,
-                history,
-                max_value,
-            } => match previous_change {
-                AggregatorChange::Data { value: prev_value } => {
-                    let new_data = merge_data_and_delta(*prev_value, delta, history, *max_value)?;
-                    *self = AggregatorChange::Data { value: new_data };
-                    Ok(())
-                },
-                AggregatorChange::StringData { .. } => Err(code_invariant_error(
-                    "Base value cannot be string for an aggregator",
-                )),
-                AggregatorChange::AggregatorDelta {
-                    delta: prev_delta,
-                    max_value: prev_max_value,
-                    history: prev_history,
-                } => {
-                    if max_value != prev_max_value {
-                        return Err(code_invariant_error(
-                            "Max value of aggregator cannot change",
-                        ));
-                    }
-                    let (new_delta, new_history) =
-                        merge_two_deltas(prev_delta, prev_history, delta, history, *max_value)?;
-                    *self = AggregatorChange::AggregatorDelta {
-                        delta: new_delta,
-                        history: new_history,
-                        max_value: *max_value,
-                    };
-                    Ok(())
-                },
-                AggregatorChange::SnapshotDelta { .. } => Err(code_invariant_error(
-                    "SnapshotDelta cannot come before AggregatorDelta",
-                )),
+    pub fn get_apply_base_id(&self, self_id: &AggregatorID) -> ApplyBase {
+        self.get_apply_base_id_option()
+            .unwrap_or(ApplyBase::Previous(*self_id))
+    }
+
+    pub fn apply_to_base(&self, base_value: AggregatorValue) -> PartialVMResult<AggregatorValue> {
+        use AggregatorApplyChange::*;
+
+        Ok(match self {
+            AggregatorDelta { delta } => {
+                AggregatorValue::Aggregator(delta.apply_to(base_value.into_aggregator_value()?)?)
             },
-            AggregatorChange::SnapshotDelta { .. } => {
-                unimplemented!()
+            SnapshotDelta { delta, .. } => {
+                AggregatorValue::Snapshot(delta.apply_to(base_value.into_aggregator_value()?)?)
```

Comment:
I'm not sure if this is a big deal. The semantics of SnapshotDelta and SnapshotDerived is that they can only be applied on base_aggregator. In this function, we are ignoring base_aggregator and applying SnapshotDelta and SnapshotDerived on any base value. Is that fine?

Reply:
these two functions go in pair - get_apply_base_id and apply_to_base

only caller knows how to fetch a value - it first needs to call get_apply_base_id to know which value to fetch, and then apply_to_base to apply it to the given value. it's on the caller to fetch correctly. 

I don't see a way to do this better/differently